File: frontend/src/types/graph.ts
/** All type definitions matching backend Pydantic models */

export type NodeType =
  | 'Technology'
  | 'Model'
  | 'Product'
  | 'AgentFramework'
  | 'AgentType'
  | 'Company'
  | 'Paper'
  | 'Benchmark';

export type RelationType =
  | 'BASED_ON'
  | 'PROPOSED_BY'
  | 'RELEASED'
  | 'COMPETES_WITH'
  | 'BELONGS_TO'
  | 'POWERS'
  | 'EVALUATED_BY'
  | 'CATEGORY_OF'
  | 'IMPROVES';

export interface GraphNode {
  id: string;
  name: string;
  node_type: NodeType;
  description: string;
  summary_zh: string;
  source_urls: string[];
  first_seen: string | null;
  last_updated: string | null;
  popularity: number;
  metadata: Record<string, unknown>;
}

export interface GraphEdge {
  source_id: string;
  target_id: string;
  relation: RelationType;
  weight: number;
  evidence: string;
}

export interface Subgraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface NodeDetail {
  node: GraphNode | null;
  incoming: GraphEdge[];
  outgoing: GraphEdge[];
  neighbor_nodes: GraphNode[];
}

export const NODE_COLORS: Record<NodeType, string> = {
  Technology: '#818cf8',
  Model: '#fbbf24',
  Product: '#34d399',
  AgentFramework: '#a78bfa',
  AgentType: '#f472b6',
  Company: '#60a5fa',
  Paper: '#94a3b8',
  Benchmark: '#f87171',
};

export const NODE_GLOWS: Record<NodeType, string> = {
  Technology: '#a5b4fc',
  Model: '#fcd34d',
  Product: '#6ee7b7',
  AgentFramework: '#c4b5fd',
  AgentType: '#f9a8d4',
  Company: '#93bbfd',
  Paper: '#cbd5e1',
  Benchmark: '#fca5a5',
};

export const NODE_LABELS: Record<NodeType, string> = {
  Technology: '技术',
  Model: '模型',
  Product: '产品',
  AgentFramework: 'Agent框架',
  AgentType: 'Agent类型',
  Company: '公司',
  Paper: '论文',
  Benchmark: '评测',
};

export const RELATION_LABELS: Record<RelationType, string> = {
  BASED_ON: '基于',
  PROPOSED_BY: '提出',
  RELEASED: '发布',
  COMPETES_WITH: '竞争',
  BELONGS_TO: '属于',
  POWERS: '驱动',
  EVALUATED_BY: '评测',
  CATEGORY_OF: '属于类别',
  IMPROVES: '改进',
};

File: backend/models/graph.py
"""Pydantic models for graph entities — all 8 node types and 9 relation types."""

import json
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Any
from datetime import datetime
from enum import Enum


class NodeType(str, Enum):
    TECHNOLOGY = "Technology"
    MODEL = "Model"
    PRODUCT = "Product"
    AGENT_FRAMEWORK = "AgentFramework"
    AGENT_TYPE = "AgentType"
    COMPANY = "Company"
    PAPER = "Paper"
    BENCHMARK = "Benchmark"


class RelationType(str, Enum):
    BASED_ON = "BASED_ON"
    PROPOSED_BY = "PROPOSED_BY"
    RELEASED = "RELEASED"
    COMPETES_WITH = "COMPETES_WITH"
    BELONGS_TO = "BELONGS_TO"
    POWERS = "POWERS"
    EVALUATED_BY = "EVALUATED_BY"
    CATEGORY_OF = "CATEGORY_OF"
    IMPROVES = "IMPROVES"


class GraphNode(BaseModel):
    """Universal node model — all 8 types share this structure."""
    id: str = Field(description="Unique identifier (slug), e.g. 'claude-opus-4-8'")
    name: str = Field(description="Display name, e.g. 'Claude Opus 4.8'")
    node_type: NodeType
    description: str = ""
    summary_zh: str = Field(default="", description="Chinese summary")
    source_urls: List[str] = Field(default_factory=list, description="URLs this node was extracted from")
    first_seen: Optional[datetime] = None
    last_updated: Optional[datetime] = None
    popularity: float = 0.0  # 0-100 score
    embedding: Optional[List[float]] = None
    metadata: dict = Field(default_factory=dict, description="Type-specific fields")

    @field_validator("metadata", mode="before")
    @classmethod
    def parse_metadata(cls, v: Any) -> dict:
        """Parse metadata from JSON string (Neo4j storage format) or pass through dict."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (json.JSONDecodeError, TypeError):
                return {}
        if isinstance(v, dict):
            return v
        return {}

    @field_validator("first_seen", "last_updated", mode="before")
    @classmethod
    def parse_neo4j_datetime(cls, v: Any) -> Optional[datetime]:
        """Convert neo4j.time.DateTime to Python datetime, or pass through."""
        if v is None:
            return None
        if isinstance(v, datetime):
            return v
        # Handle neo4j.time.DateTime (has to_native / fromtimestamp methods)
        if hasattr(v, "to_native"):
            return v.to_native()
        if hasattr(v, "isoformat"):
            return datetime.fromisoformat(v.isoformat().replace("Z", "+00:00"))
        return v


class GraphEdge(BaseModel):
    """Universal relationship model."""
    source_id: str
    target_id: str
    relation: RelationType
    weight: float = 1.0
    evidence: str = Field(default="", description="Sentence or source that justifies this edge")


class Subgraph(BaseModel):
    """A slice of the graph returned for visualization."""
    nodes: List[GraphNode]
    edges: List[GraphEdge]


class NodeDetail(BaseModel):
    """Full detail for the side panel — node + its immediate neighborhood."""
    node: Optional[GraphNode] = None
    incoming: List[GraphEdge] = Field(default_factory=list)
    outgoing: List[GraphEdge] = Field(default_factory=list)
    neighbor_nodes: List[GraphNode] = Field(default_factory=list)
